File: micro-dummies/services/suppliers-ms/internal/core/domain/events.js
// Domain events for Supplier
class SupplierEvent {
    constructor(eventType, data) {
        this.eventType = eventType;
        this.data = data;
        this.timestamp = new Date().toISOString();
        this.eventId = this.generateEventId();
    }

    generateEventId() {
        return `${Date.now()}-${Math.random().toString(36).substr(2, 9)}`;
    }

    toJSON() {
        return {
            eventId: this.eventId,
            eventType: this.eventType,
            data: this.data,
            timestamp: this.timestamp
        };
    }
}

class SupplierCreatedEvent extends SupplierEvent {
    constructor(supplier) {
        super('supplier.created', supplier);
    }
}

class SupplierUpdatedEvent extends SupplierEvent {
    constructor(supplier) {
        super('supplier.updated', supplier);
    }
}

class MedicineEvent {
    constructor(eventType, data) {
        this.eventType = eventType;
        this.data = data;
        this.timestamp = new Date().toISOString();
        this.eventId = this.generateEventId();
    }

    generateEventId() {
        return `${Date.now()}-${Math.random().toString(36).substr(2, 9)}`;
    }

    toJSON() {
        return {
            eventId: this.eventId,
            eventType: this.eventType,
            data: this.data,
            timestamp: this.timestamp
        };
    }
}

class MedicineCreatedEvent extends MedicineEvent {
    constructor(medicine) {
        super('medicine.updated', medicine);
    }
}

class MedicineUpdatedEvent extends MedicineEvent {
    constructor(medicine) {
        super('medicine.updated', medicine);
    }
}

module.exports = {
    SupplierEvent,
    SupplierCreatedEvent,
    SupplierUpdatedEvent,
    MedicineEvent,
    MedicineCreatedEvent,
    MedicineUpdatedEvent
};

File: procurement-supply/suppliers/internal/core/domain/events.go
package domain

import (
	"time"
)

type EventType string

type Event[T any] struct {
	EventType `json:"event_type"`
	Data      T         `json:"data"`
	Timestamp time.Time `json:"timestamp"`
}

const (
	MedicineCreatedEvent EventType = "medicine.created"
	MedicineUpdatedEvent EventType = "medicine.updated"
	MedicineDeletedEvent EventType = "medicine.deleted"
)

// NewMedicineEvent creates a new medicine event with validation
func NewMedicineEvent(data Medicine, event EventType) *Event[Medicine] {
	return &Event[Medicine]{
		EventType: event,
		Data:      data,
		Timestamp: time.Now(),
	}
}
